marketplace_standard_app_api/main.py

```diff
@@ -108,29 +108,155 @@ async def heartbeat() -> HTMLResponse:
     return HTMLResponse(content="<html><body>OK</body></html>", status_code=200)
 
 
-@api.post(
+@api.get(
     "/datasets",
+    operation_id="listDatasets",
+    tags=["DataSource"],
+    response_model=DatasetListResponse,
+    responses={
+        204: {"description": "No datasets found."},
+        401: {"description": "Not authenticated."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+)
+async def list_datasets(limit: int = 100, offset: int = 0) -> DatasetListResponse:
+    """List all datasets."""
+    raise HTTPException(status_code=501, detail="Not implemented.")
+
+
+CREATE_DATASET_DESCRIPTION = """
+To add custom metadata, add keys to the header of the form:
+
+- X-Object-Metadata-name: value
+
+Where 'name' is the name of the metadata key and 'value' is the
+corresponding value.
+
+Note: This operation is in compliance with the OpenStack Swift object
+storage API:
+https://docs.openstack.org/api-ref/object-store/index.html#create-or-replace-object
+"""
+
+
+@api.put(
+    "/datasets/{dataset_id}",
+    name="Create or Replace Dataset",
+    operation_id="createOrReplaceDataset",
+    tags=["DataSink"],
+    response_model=DatasetCreateResponse,
+    status_code=201,
+    responses={
+        401: {"description": "Not authenticated."},
+        500: {"description": "Internal server error."},
+        501: {"description": "Not implemented."},
+        503: {"description": "Service unavailable."},
+    },
+    description="Create or replace a dataset.\n" + CREATE_DATASET_DESCRIPTION,
+)
+@api.put(
+    "/datasets/",
     operation_id="createDataset",
     tags=["DataSink"],
     response_model=DatasetCreateResponse,
+    status_code=201,
     responses={
         401: {"description": "Not authenticated."},
         500: {"description": "Internal server error."},
         501: {"description": "Not implemented."},
         503: {"description": "Service unavailable."},
     },
+    description="Create a dataset.\n" + CREATE_DATASET_DESCRIPTION,
 )
-async def create_dataset(dataset: DatasetModel) -> DatasetCreateResponse:
-    """Create a new dataset."""
+async def create_dataset(
+    file: UploadFile,
+    dataset_id: Optional[DatasetId] = None,
+) -> DatasetCreateResponse:
+    """Create a new or replace an existing dataset."""
+    raise HTTPException(status_code=501, detail="Not implemented.")
+
+
+@api.post(
+    "/datasets/",
+    operation_id="createDatasetMetadata",
+    name="Create Dataset Metadata",
+    tags=["DataSink"],
+    status_code=202,
```

Comment:
Why not use `201` for these creations as well?

Reply:
To my understanding that's because the primary resource in this context is the _object_, which is not created via this operation. I would advise against deviating from the [SWIFT or S3 specification](https://docs.openstack.org/api-ref/object-store/index.html?expanded=create-or-update-object-metadata-detail).